### PR TITLE
Fix AAIOB when using select star with natural join [HZ-1329]

### LINKED ISSUE #21325: AIOOB when expanding star in JOIN with USING
The following test:
```java
public class FooTest extends SqlTestSupport {

    @Test
    public void test() {
        HazelcastInstance inst = createHazelcastInstance();
        createMapping(inst, "Hzl_Transaction", Long.class, Transaction.class);
        createMapping(inst, "Hzl_Account", Long.class, Account.class);

        inst.getSql().execute("select * from Hzl_Transaction JOIN Hzl_Account USING(bankId, accountNumber)");
    }

    public static class Transaction implements Serializable {
        public String bankId, accountNumber;
        public int amount;
    }

    public static class Account implements Serializable {
        public String bankId, accountNumber;
        public int balance;
    }
}
```
fails with:
```
com.hazelcast.sql.impl.QueryException: Index 8 out of bounds for length 8
	at com.hazelcast.sql.impl.QueryException.error(QueryException.java:65)
	at com.hazelcast.jet.sql.impl.parse.QueryParser.parse(QueryParser.java:67)
	at com.hazelcast.jet.sql.impl.OptimizerContext.parse(OptimizerContext.java:151)
	at com.hazelcast.jet.sql.impl.CalciteSqlOptimizer.prepare(CalciteSqlOptimizer.java:236)
	at com.hazelcast.sql.impl.SqlServiceImpl.prepare(SqlServiceImpl.java:252)
	at com.hazelcast.sql.impl.SqlServiceImpl.query0(SqlServiceImpl.java:236)
	at com.hazelcast.sql.impl.SqlServiceImpl.execute(SqlServiceImpl.java:193)
	... 17 more
Caused by: java.lang.ArrayIndexOutOfBoundsException: Index 8 out of bounds for length 8
	at com.google.common.collect.RegularImmutableList.get(RegularImmutableList.java:75)
	at org.apache.calcite.sql.validate.SqlValidatorImpl$Permute.permute(SqlValidatorImpl.java:7106)
	at org.apache.calcite.sql.validate.SqlValidatorImpl.expandStar(SqlValidatorImpl.java:664)
	at org.apache.calcite.sql.validate.SqlValidatorImpl.expandSelectItem(SqlValidatorImpl.java:426)
	at org.apache.calcite.sql.validate.SqlValidatorImpl.validateSelectList(SqlValidatorImpl.java:4409)
	at org.apache.calcite.sql.validate.SqlValidatorImpl.validateSelect(SqlValidatorImpl.java:3652)
	at org.apache.calcite.sql.validate.SelectNamespace.validateImpl(SelectNamespace.java:64)
	at org.apache.calcite.sql.validate.AbstractNamespace.validate(AbstractNamespace.java:89)
	at org.apache.calcite.sql.validate.SqlValidatorImpl.validateNamespace(SqlValidatorImpl.java:1100)
	at org.apache.calcite.sql.validate.SqlValidatorImpl.validateQuery(SqlValidatorImpl.java:1071)
	at com.hazelcast.jet.sql.impl.validate.HazelcastSqlValidator.validateQuery(HazelcastSqlValidator.java:170)
	at org.apache.calcite.sql.SqlSelect.validate(SqlSelect.java:247)
	at org.apache.calcite.sql.validate.SqlValidatorImpl.validateScopedExpression(SqlValidatorImpl.java:1046)
	at org.apache.calcite.sql.validate.SqlValidatorImpl.validate(SqlValidatorImpl.java:752)
	at com.hazelcast.jet.sql.impl.validate.HazelcastSqlValidator.validate(HazelcastSqlValidator.java:165)
	at com.hazelcast.jet.sql.impl.parse.QueryParser.parse0(QueryParser.java:80)
	at com.hazelcast.jet.sql.impl.parse.QueryParser.parse(QueryParser.java:56)
	... 22 more
```

### PR DESCRIPTION
**Pull request by hakanmemisoglu**
Permute method applied in Natural Join or Using Join causes an exception because the method doesn't handle
our hidden fields. It is a private method and cannot be intercepted safely, therefore a copy of SqlValidator
is put under the sources and the condition for triggering the permute operation is removed.

Checklist:
- [x] Labels (`Team:`, `Type:`, `Source:`, `Module:`) and Milestone set
- [x] Label `Add to Release Notes` or `Not Release Notes content` set
- [x] Request reviewers if possible

Fix #21325

### CODE REVIEW

**Comment by krzysztofslusarski** [comment]
The real thing in that PR is not shown by GitHub. We copy the ```SqlValidatorImpl``` to our codebase and we are changing a method ```expandStar``` by removing:

```java
      // If NATURAL JOIN or USING is present, move key fields to the front of
      // the list, per standard SQL. Disabled if there are dynamic fields.
      if (!hasDynamicStruct || Bug.CALCITE_2400_FIXED) {
        SqlNode from = requireNonNull(scope.getNode().getFrom(),
            () -> "getFrom for " + scope.getNode());
        new Permute(from, 0).permute(selectItems, fields);
      }
```

**Review by krzysztofslusarski — APPROVED** [review]

**Comment by krzysztofslusarski** [comment]
@hakanmemisoglu new calcite has just been published. I believe we need to upgrade to 1.31 and copy new validator.

**Review by krzysztofslusarski — CHANGES_REQUESTED** [review]
To recheck after upgrading calcite

**Review by krzysztofslusarski — APPROVED** [review]

File: hazelcast-sql/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java [hunk 1]
```diff
@@ -0,0 +1,7271 @@
...
+                // Commenting out this section:
+                // https://github.com/hazelcast/hazelcast/issues/21325
+                //
+                // If NATURAL JOIN or USING is present, move key fields to the front of
+                // the list, per standard SQL. Disabled if there are dynamic fields.
+                // if (!hasDynamicStruct || Bug.CALCITE_2400_FIXED) {
+                //    SqlNode from = requireNonNull(scope.getNode().getFrom(),
+                //            () -> "getFrom for " + scope.getNode());
+                //    new Permute(from, 0).permute(selectItems, fields);
+                // }
```
Comment (viliam-durina) [inline hazelcast-sql/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java:687]: With this you essentially broke this type of join. For example, if you have `t1(a, b)` and `t2(c, b)` and you do `FROM t1 JOIN t2 USING (b)` or `FROM t1 NATURAL JOIN t2`, the query is supposed to have these columns on the output: `b, a, c`. With this code commented out, it has `a, b, c, b` - `b` is there twice with the same value, and it should be at the beginning, as the calcite code comment says

**Review by viliam-durina — CHANGES_REQUESTED** [review]

**Comment by viliam-durina** [comment]
I propose to close this PR for now. It replaces an error with an incorrect result. The workaround is to not use `SELECT *` in combination with `USING` or `NATURAL JOIN`. 

Hidden fields aren't natively supported by calcite. It is our implementation of hidden fields that causes it.

**Pull request closed without merging**